File: java/backtracking/combination_sum.java
// beats 99% of Java users
class Solution 
{
    public List<List<Integer>> combSum;
    public List<List<Integer>> combinationSum(int[] candidates, int target) 
    {
        combSum = new ArrayList<>();
        backtr(candidates, target, new ArrayList<Integer>(), 0);
        return combSum;
    }
    public void backtr (int[] nums, int target, List<Integer> cur, int index)
    {
        for ( int i=index; i < nums.length; i++)
        {
            if ( nums[i] < target )
            {
                cur.add(nums[i]);
                backtr(nums, target-nums[i],cur, i);
                cur.remove(cur.size()-1);
            }
            else if ( nums[i] == target )
            {
                List<Integer> member = new ArrayList<Integer>();
                cur.add(nums[i]);
                for ( int j=0; j<cur.size(); j++)
                {
                    member.add(cur.get(j));
                }
                combSum.add(member);
                cur.remove(cur.size()-1);
            }
        }
    }
}

File: python/backtracking/combination_sum.py
import copy
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        temp = []

        def back_track(i, cur, cur_sum):
            if cur_sum == target:
                cp = copy.deepcopy(cur)
                res.append(cp)
                return
            if cur_sum > target or i >= len(candidates):
                return
            
            cur.append(candidates[i])
            back_track(i, cur, cur_sum+candidates[i])
            cur.pop()
            back_track(i+1, cur, cur_sum)

        back_track(0, temp, 0)

        return res